### PR TITLE
Fix parsing many test suites on file

### PR DESCRIPTION
**Pull request by ETtestim**
Instead of assuming we get only one test suite in the file, now we parse all suites on file

### CODE REVIEW

File: src/junitParser.ts [hunk 1]
```diff
@@ -70,26 +70,36 @@ async function parseFile(file: string, projectTokenDictionaryStrs: string[]) {
...
     const parser = new XMLParser({ allowBooleanAttributes: true, ignoreAttributes: false, attributeNamePrefix: '' });
     const report = parser.parse(data) as Partial<JUnitReport>;
-
-    return parseSuite(report, file, projectTokenDictionaryStrs);
+    const testsuites = castArray(report.testsuites?.testsuite);
+    const result: InternalTestResult[] = [];
+    for (const testsuite of testsuites) {
+        await parseSuite(file, projectTokenDictionaryStrs, testsuite as JUnitSuite, result);
+    }
+    return result;
```
Comment (atlowChemi) [inline src/junitParser.ts]: Why pass the `result` array instead of mapping here and then not having to change `parseSuite`?
```suggestion
    return Promise.all(testsuites.map(testsuite => parseSuite(file, projectTokenDictionaryStrs, testsuite as JUnitSuite));
```

Comment (ETtestim) [reply src/junitParser.ts]: When you are right you are right

File: src/junitParser.ts [hunk 2]
```diff
@@ -70,18 +70,17 @@ async function parseFile(file: string, projectTokenDictionaryStrs: string[]) {
     const data: string = fs.readFileSync(file, 'utf8');
     const parser = new XMLParser({ allowBooleanAttributes: true, ignoreAttributes: false, attributeNamePrefix: '' });
     const report = parser.parse(data) as Partial<JUnitReport>;
-
-    return parseSuite(report, file, projectTokenDictionaryStrs);
+    const testSuites = castArray(report.testsuites?.testsuite);
+    return Promise.all((testSuites as JUnitSuite[]).map(async (testSuite: JUnitSuite) => parseSuite(file, projectTokenDictionaryStrs, testSuite)));
```
Comment (atlowChemi) [inline src/junitParser.ts]: Shouldn't these types all be infered?

```suggestion
    const testSuites = castArray(report.testsuites?.testsuite || []);
    return Promise.all((testSuites).map(async (testSuite) => parseSuite(file, projectTokenDictionaryStrs, testSuite)));
```

File: src/junitParser.ts [hunk 3]
```diff
@@ -212,3 +213,11 @@ export async function getTestReports(inputs: Readonly<ReturnType<typeof parseInp
...
     };
 }
+// parseTestReports('elad', 'summary', '/Users/eladtal/testim-junit-reporter/testim-junit-reporter-1/src/report-small_res_ui_test_save_delete.xml', [
+//     'testimProjectToken:token',
+//     'testimProjectToken2:token2',
+// ]);
+parseTestReports('elad', 'summary', '/Users/eladtal/testim-junit-reporter/testim-junit-reporter-1/src/report-test_plan_sanity_params_test_plan_basic_and_basic_selenium.xml', [
+    'testimProjectToken:token',
+    'testimProjectToken2:token2',
+]);
```
Comment (atlowChemi) [inline src/junitParser.ts]:
```suggestion

```

File: src/junitParser.ts [hunk 4]
```diff
@@ -60,8 +60,7 @@ interface JUnitSuite {
     testcase: JUnitTestCase | JUnitTestCase[];
 }
 interface JUnitReport {
-    testsuite?: JUnitSuite;
-    testsuites?: { testsuite?: JUnitSuite };
+    testsuites?: { testsuite?: JUnitSuite | JUnitSuite[] };
```
Comment (atlowChemi) [inline src/junitParser.ts:64]: We should still keep `testsuite?: JUnitSuite;` to support other `junit` files as well (checkout the smtp-server)

File: src/junitParser.ts [hunk 5]
```diff
@@ -70,18 +69,17 @@ async function parseFile(file: string, projectTokenDictionaryStrs: string[]) {
     const data: string = fs.readFileSync(file, 'utf8');
     const parser = new XMLParser({ allowBooleanAttributes: true, ignoreAttributes: false, attributeNamePrefix: '' });
     const report = parser.parse(data) as Partial<JUnitReport>;
-
-    return parseSuite(report, file, projectTokenDictionaryStrs);
+    const testSuites = castArray(report.testsuites?.testsuite || []);
```
Comment (atlowChemi) [inline src/junitParser.ts]:
```suggestion
    const testSuites = report.testsuite || castArray(report.testsuites?.testsuite || []);
```

File: src/junitParser.ts [hunk 6]
```diff
@@ -70,18 +69,17 @@ async function parseFile(file: string, projectTokenDictionaryStrs: string[]) {
...
+    return Promise.all(testSuites.map(async testSuite => parseSuite(file, projectTokenDictionaryStrs, testSuite)));
 }
 
-async function parseSuite(report: JUnitReport, fileName: string, projectTokenDictionaryStrs: string[]) {
-    const testsuite = report.testsuite || report.testsuites?.testsuite;
-    const result: InternalTestResult = { fileName, name: testsuite?.name, totalCount: 0, skipped: 0, failedEvaluating: parseInt(testsuite?.['failure-evaluating'] || '0') || 0, annotations: [] };
-    if (!testsuite?.testcase) {
+async function parseSuite(fileName: string, projectTokenDictionaryStrs: string[], testSuite: JUnitSuite) {
```
Comment (atlowChemi) [inline src/junitParser.ts]: can we keep `testSuite` as `testsuite` to avoid unrelated changes?

File: src/junitParser.ts [hunk 7]
```diff
@@ -123,24 +121,26 @@ async function parseTestReports(checkName: string, summary: string, reportPathsG
     const globber = await glob.create(reportPathsGlob);
     for await (const file of globber.globGenerator()) {
         core.info(`Parsing report file: ${file}`);
-
-        const { totalCount, skipped, annotations, failedEvaluating, name, fileName } = await parseFile(file, projectTokenDictionaryStrs);
-        if (totalCount === 0) {
-            continue;
+        const suites = await parseFile(file, projectTokenDictionaryStrs);
+        for (const suite of suites) {
```
Comment (atlowChemi) [inline src/junitParser.ts:126]: maybe this could later be re-written as an async generator, but for now - this works for me

File: src/junitParser.ts (hunk 6 above)
Comment (ETtestim) [reply src/junitParser.ts]: NP

File: src/junitParser.ts [hunk 8]
```diff
@@ -70,12 +70,11 @@ async function parseFile(file: string, projectTokenDictionaryStrs: string[]) {
     const data: string = fs.readFileSync(file, 'utf8');
     const parser = new XMLParser({ allowBooleanAttributes: true, ignoreAttributes: false, attributeNamePrefix: '' });
     const report = parser.parse(data) as Partial<JUnitReport>;
-
-    return parseSuite(report, file, projectTokenDictionaryStrs);
+    const testsuites = report.testsuite ? castArray(report.testsuite) : castArray(report.testsuites?.testsuite);
```
Comment (atlowChemi) [inline src/junitParser.ts]:
```suggestion
    const testsuites = castArray(report.testsuite || report.testsuites?.testsuite || []);
```

**Comment by barspielberg** [comment]
@ETtestim Can you add some unit tests?

**Comment by ETtestim** [comment]
> @ETtestim Can you add some unit tests?

**Comment by ETtestim** [comment]
@atlowChemi @barspielberg You want to add vitest or some other 3 party for testing?

**Comment by atlowChemi** [comment]
> @atlowChemi @barspielberg You want to add vitest or some other 3 party for testing?

I think for this simple use case there is no reason not to use the built-in node test-runner.

@barspielberg Thanks for that awesome comment 👍🏽

**Review by atlowChemi — DISMISSED** [review]
as per to @barspielberg's comment regarding missing UT

File: src/test/junitParser.test.ts
Comment (barspielberg) [inline src/test/junitParser.test.ts:1]: 🥇

File: package.json [hunk 9]
```diff
@@ -12,7 +12,8 @@
         "format": "prettier --write **/*.ts",
         "format-check": "prettier --check **/*.ts",
         "lint": "eslint --ext .ts src/",
-        "all": "yarn type-check && yarn build && yarn format && yarn lint"
+        "all": "yarn type-check && yarn build && yarn format && yarn lint",
+        "test": "node --test --no-warnings=ExperimentalWarning --loader ts-node/esm src/test/*.test.ts"
```
Comment (atlowChemi) [inline package.json]: Can we use `tsx` instead? it would be faster and closer to the code we build (as it uses `esbuild`)

```suggestion
        "test": "node --test --no-warnings=ExperimentalWarning --import tsx src/test/*.test.ts"
```

Also, I think `--no-warnings=ExperimentalWarning` can now be removed as the CI now runs on Node v22

File: tsconfig.json [hunk 10]
```diff
@@ -11,5 +11,7 @@
         "esModuleInterop": true,
         "skipLibCheck": true,
     },
-    "exclude": ["node_modules", "**/*.test.ts"]
-}
+    "exclude": [
+        "node_modules"
+    ]
```
Comment (atlowChemi) [inline tsconfig.json:18]: Could you ensure this doesn't affect the files outputted during `yarn build`?

File: src/test/junitParser.test.ts [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+            message: 'Step Failed: Element is not visible More info at: \nhttps://app.testim.io/#/project/coco/branch/master/test/1234567B?result-id=1234567B',
+            raw_details: 'test - first suite second test:\n(Step Failed: Element is not visible More info at: https://app.testim.io/#/project/coco/branch/master/test/1234567B?result-id=1234567B)',
+            path: 'https://app.testim.io/#/project/coco/branch/master/test/1234567B?result-id=1234567B',
+            end_line: 1,
+            start_line: 1,
+        },
+    ],
+    failed: 1,
+    passed: 0,
+});
```
Comment (atlowChemi) [inline src/test/junitParser.test.ts]: Could we simplify this with the test-runner snapshots?
See https://nodejs.org/docs/latest/api/test.html#snapshot-testing

File: package.json (hunk 9 above)
Comment (atlowChemi) [reply package.json]: Also, the github workflow should include this?

File: tsconfig.json (hunk 10 above)
Comment (ETtestim) [reply tsconfig.json:18]: image
run the one with the change and other without same sizes and also looked inside the index.js but didn't find anything

File: src/test/junitParser.test.ts (hunk 11 above)
Comment (ETtestim) [reply src/test/junitParser.test.ts:39]: Do you have a project that uses this?
I couldn't find how to use it and the docs example doesn't work for me

File: src/junitParser.ts (hunk 8 above)
Comment (atlowChemi) [inline src/junitParser.ts]:
```suggestion
    const testsuites = castArray(report.testsuite || report.testsuites?.testsuite);
```

File: src/test/junitParser.test.ts [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+//         const testReports = await parseTestReports('test', 'test', `${DIRNAME}/xml/testsuites.xml`, []);
+//         assert.strictEqual(testReports.length, 3);
+//         assert.deepEqual(testReports[0], getFirstSuite('xml/testsuites.xml'));
+//     });
+//     it('should parse junit report with 1 suite', async () => {
+//         const testReports = await parseTestReports('test', 'test', `${DIRNAME}/xml/testsuite.xml`, []);
+//         assert.strictEqual(testReports.length, 1);
+//         assert.deepEqual(testReports[0], getFirstSuite('xml/testsuite.xml'));
+//     });
+// });
```
Comment (atlowChemi) [inline src/test/junitParser.test.ts]: Can this be removed?

File: src/test/junitParser.test.ts [hunk 13]
```diff
@@ -0,0 +1,29 @@
+import { test } from 'node:test';
+import { parseTestReports } from '../junitParser';
+
+const DIRNAME = __dirname;
```
Comment (atlowChemi) [inline src/test/junitParser.test.ts]: Why not `import.meta.dirname`?

Comment (ETtestim) [reply src/test/junitParser.test.ts]: The 'import.meta' meta-property is not allowed in files which will build into CommonJS output.
this is the error I get once trying to use `import.meta.dirname`

Comment (ETtestim) [reply src/test/junitParser.test.ts]: @atlowChemi

Comment (atlowChemi) [reply src/test/junitParser.test.ts]: Oh so let's transpile to ESM. 
Coulsd you add `"type": "module"` to the `package.json`?

File: src/junitParser.ts [hunk 14]
```diff
@@ -70,12 +70,11 @@ async function parseFile(file: string, projectTokenDictionaryStrs: string[]) {
     const data: string = fs.readFileSync(file, 'utf8');
     const parser = new XMLParser({ allowBooleanAttributes: true, ignoreAttributes: false, attributeNamePrefix: '' });
     const report = parser.parse(data) as Partial<JUnitReport>;
-
-    return parseSuite(report, file, projectTokenDictionaryStrs);
+    const testsuites = castArray(report?.testsuite || report.testsuites?.testsuite);
```
Comment (atlowChemi) [inline src/junitParser.ts]: Can report be undefined?
I don't think it can, so simplify to:

```suggestion
    const testsuites = castArray(report.testsuite || report.testsuites?.testsuite);
```

If it can:

```suggestion
    const testsuites = castArray(report?.testsuite || report?.testsuites?.testsuite);
```

File: src/test/junitParser.test.ts (hunk 13 above)
Comment (ETtestim) [reply src/test/junitParser.test.ts]: this starts a regression
image
Which I'm not able to get out.
Even after trying to change the moduleResolution it doesn't work.

Comment (ETtestim) [reply src/test/junitParser.test.ts]: @atlowChemi Tried again to find solution but couldn't find one its annoying

File: .github/workflows/build.yaml [hunk 15]
```diff
@@ -27,3 +27,5 @@ jobs:
                 run: yarn format-check
             -   name: Lint
                 run: yarn lint
+            -   name: Unit Test
```
Comment (atlowChemi) [inline .github/workflows/build.yaml]:
```suggestion
            -   name: Unit Tests
```

File: src/junitParser.ts [hunk 16]
```diff
@@ -70,12 +70,11 @@ async function parseFile(file: string, projectTokenDictionaryStrs: string[]) {
     const data: string = fs.readFileSync(file, 'utf8');
     const parser = new XMLParser({ allowBooleanAttributes: true, ignoreAttributes: false, attributeNamePrefix: '' });
     const report = parser.parse(data) as Partial<JUnitReport>;
-
-    return parseSuite(report, file, projectTokenDictionaryStrs);
+    const testsuites = castArray(report?.testsuite || report?.testsuites?.testsuite);
```
Comment (atlowChemi) [inline src/junitParser.ts]: Before the changes we did not handle `report` as nullable, why is it now?
```suggestion
    const testsuites = castArray(report.testsuite || report.testsuites?.testsuite);
```

File: src/junitParser.ts [hunk 17]
```diff
@@ -116,31 +115,33 @@ async function parseSuite(report: JUnitReport, fileName: string, projectTokenDic
...
-        const { totalCount, skipped, annotations, failedEvaluating, name, fileName } = await parseFile(file, projectTokenDictionaryStrs);
-        if (totalCount === 0) {
-            continue;
+        const suites = await parseFile(file, projectTokenDictionaryStrs);
+        for (const suite of suites) {
+            const { totalCount, skipped, annotations, failedEvaluating, name, fileName } = suite;
+            if (totalCount === 0) {
+                continue;
+            }
+            const failed = annotations.filter((an: { annotation_level: string; }) => an.annotation_level !== 'notice').length;
```
Comment (atlowChemi) [inline src/junitParser.ts]: Why is this change required?
This should be typed already from the return type?

File: src/junitParser.ts [hunk 18]
```diff
@@ -155,7 +156,7 @@ export async function getTestReports(inputs: Readonly<ReturnType<typeof parseInp
     core.info(`Retrieved ${reportsCount} report globs/files to process.`);
 
     const allResults = await Promise.all(
-        inputs.reportPaths.map(async (_, i) =>
+        inputs.reportPaths.map(async (_: any, i: any) =>
```
Comment (atlowChemi) [inline src/junitParser.ts]: Why?
`inputs` is typed, which means `inputs.reportPaths` is also typed

File: src/test/junitParser.test.ts (hunk 13 above)
Comment (atlowChemi) [reply src/test/junitParser.test.ts]: @ETtestim this means you need to add file extension, IIRC

File: src/test/junitParser.test.ts [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+
+test('should parse junit report with 3 suites', async (t: any) => {
```
Comment (atlowChemi) [inline src/test/junitParser.test.ts]: why `any`?

File: src/test/junitParser.test.ts [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+    const testReports = await parseTestReports('test', 'test', `${DIRNAME}/xml/testsuites.xml`, []);
+    t.assert.snapshot(testReports.length);
```
Comment (atlowChemi) [inline src/test/junitParser.test.ts]: Why is this a snapshot assertion?
```suggestion
    t.assert.strictEqual(testReports.length, 3);
```

File: src/test/junitParser.test.ts [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+    t.assert.snapshot(testReports[0].fileName);
+    t.assert.snapshot(testReports[0].annotations.length);
+});
+test('should parse junit report with 1 suite', async (t: any) => {
+    const testReports = await parseTestReports('test', 'test', `${DIRNAME}/xml/testsuite.xml`, []);
+    t.assert.snapshot(testReports.length);
```
Comment (atlowChemi) [inline src/test/junitParser.test.ts]: Why is this a snapshot assertion?
```suggestion
    t.assert.strictEqual(testReports.length, 1);
```

File: src/test/junitParser.test.ts
Comment (atlowChemi) [inline src/test/junitParser.test.ts:1]: Could you add a test ensuring we only attempt calling Testim API when key provided for given Project ID?
Also a test validating the annotations, and not only the file name

File: src/test/junitParser.test.ts.snapshot [hunk 22]
```diff
@@ -0,0 +1,23 @@
+exports[`should parse junit report with 1 suite 1`] = `
+1
+`;
+
+exports[`should parse junit report with 1 suite 2`] = `
+"/Users/eladtal/testim-junit-reporter/testim-junit-reporter-1/src/test/xml/testsuite.xml"
```
Comment (atlowChemi) [inline src/test/junitParser.test.ts.snapshot]: This will fail since the CI doesn't run on your laptop?

File: src/test/junitParser.test.ts (hunk 19 above)
Comment (ETtestim) [reply src/test/junitParser.test.ts]: when I let it be TestContext it doesn't like it.
Property 'assert' does not exist on type 'TestContext'

File: src/test/junitParser.test.ts (hunk 13 above)
Comment (atlowChemi) [reply src/test/junitParser.test.ts]: @ETtestim  try now

File: src/test/junitParser.test.ts
Comment (ETtestim) [reply src/test/junitParser.test.ts:1]: I will have another PR for this repo next sprint can we add those test then?

File: src/junitParser.ts (hunk 14 above)
Comment (atlowChemi) [inline src/junitParser.ts]: Still no reason for optional chaining AFAICT?
```suggestion
    const testsuites = castArray(report.testsuite || report.testsuites?.testsuite);
```

File: src/junitParser.ts [hunk 23]
```diff
@@ -70,12 +70,11 @@ async function parseFile(file: string, projectTokenDictionaryStrs: string[]) {
     const data: string = fs.readFileSync(file, 'utf8');
     const parser = new XMLParser({ allowBooleanAttributes: true, ignoreAttributes: false, attributeNamePrefix: '' });
     const report = parser.parse(data) as Partial<JUnitReport>;
-
-    return parseSuite(report, file, projectTokenDictionaryStrs);
+    const testsuites = castArray(report?.testsuite || report.testsuites?.testsuite);
+    return Promise.all(testsuites.map(async (testsuite: JUnitSuite | undefined) => parseSuite(file, projectTokenDictionaryStrs, testsuite)));
```
Comment (atlowChemi) [inline src/junitParser.ts]: Why `: JUnitSuite | undefined`? What is it typed as?

File: src/test/junitParser.test.ts (hunk 19 above)
Comment (atlowChemi) [reply src/test/junitParser.test.ts]: Can you replace the `any` here with `// @ts-expect-error` when calling assert so once `@types/node` has this added we can get types again?

File: src/test/junitParser.test.ts [hunk 24]
```diff
@@ -0,0 +1,15 @@
+import { test } from 'node:test';
+import { parseTestReports } from '../junitParser.ts';
+
+const DIRNAME = import.meta.dirname;
+
+test('should parse junit report with 3 suites', async (t: any) => {
```
Comment (atlowChemi) [inline src/test/junitParser.test.ts]:
```suggestion
test('should parse junit report with multiple suites', async (t: any) => {
```

File: src/test/junitParser.test.ts
Comment (atlowChemi) [reply src/test/junitParser.test.ts:1]: What PR? which sprint?
I am confused

File: src/test/junitParser.test.ts [hunk 25]
```diff
@@ -0,0 +1,18 @@
+import { test } from 'node:test';
+import { parseTestReports } from '../junitParser.ts';
+
+const DIRNAME = import.meta.dirname;
+test('should parse junit report with multiple suites', async (t) => {
+    const testReports = await parseTestReports('test', 'test', `${DIRNAME}/xml/testsuites.xml`, []);
+    // @ts-expect-error
+    t.assert.strictEqual(testReports.length, 3);
+    // @ts-expect-error
+    t.assert.strictEqual(testReports[0].annotations.length, 2);
```
Comment (atlowChemi) [inline src/test/junitParser.test.ts:10]: we could also just use `assert` directly and avoid `// @ts-expect-error`

Comment (atlowChemi) [inline src/test/junitParser.test.ts:10]: Could we snapshot test the entire `testReports` object?

**Comment by ETtestim** [comment]
The code inside here was copied inside Testim code.
With the contribute file!
As I'm leaving Testim I'm closing this PR
